clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+    return SR::Top;
+
+  if (Val == &Env.getBoolLiteralValue(true))
+    return SR::True;
+  
+  if (Val == &Env.getBoolLiteralValue(false))
+    return SR::False;
+
+  return SR::Unknown;
+}
+
+// Computes satisfiability by using the flow condition. Slower, but more precise.
+SatisfiabilityResult computeSatisfiability(BoolValue *Val, const Environment &Env) {
+  // Early return with the fast compute values.
+  if (SatisfiabilityResult ShallowResult = shallowComputeSatisfiability(Val, Env);
+      ShallowResult != SR::Unknown) {
+    return ShallowResult;
+  }
+
+  if (Env.flowConditionImplies(Val->formula()))
+    return SR::True;
+  
+  if (Env.flowConditionImplies(Env.arena().makeNot(Val->formula())))
+    return SR::False;
+
+  return SR::Unknown;
+}
+
+inline BoolValue &getVal(llvm::StringRef Name, Value &RootValue) {
+    auto *Value = cast_or_null<BoolValue>(RootValue.getProperty(Name));
+    assert(Value && "boolean value was not created");
+    return *Value;
+}
+
+void initializeRootValue(Value &RootValue, Environment &Env) {
+
+    Arena &A = Env.arena();
+
+    BoolValue &IsNull = A.makeAtomValue();
+    BoolValue &IsNonnull = A.makeAtomValue();
+    RootValue.setProperty(kIsNull, IsNull);
+    RootValue.setProperty(kIsNonnull, IsNonnull);
+
+    Env.addToFlowCondition(A.makeOr(IsNull.formula(), IsNonnull.formula()));
+}
+
+void setLValue(const Expr &E, Value &Val, Environment &Env) {
+  StorageLocation *SL = Env.getStorageLocation(E);
+  if (!SL) {
+    SL = &Env.createStorageLocation(E);
+    Env.setStorageLocation(E, *SL);
+  }
+
+  Env.setValue(*SL, Val);
+}
+
+void setUnknownValue(const Expr &E, Value &Val, Environment &Env) {
+  if (E.isLValue()) {
+    setLValue(E, Val, Env);
+  } else if (E.isPRValue()) {
+    Env.setValue(E, Val);
+  } else {
+    llvm_unreachable("all value cases covered");
+  }
+}
+
+Value *getValue(const Expr &Var, Environment &Env) {
+    if (auto *EnvVal = Env.getValue(Var)) {
+        // FIXME: Hack
+        if (!EnvVal->getProperty(kIsNull) && !EnvVal->getProperty(kIsNonnull)) {
+            initializeRootValue(*EnvVal, Env);
+        }
+
+        return EnvVal;
+    }
+
+    auto *RootValue = Env.createValue(Var.getType());
+
+    if (!RootValue) 
+        return nullptr;
+
+    initializeRootValue(*RootValue, Env);
+
+    setLValue(Var, *RootValue, Env);
+
+    return RootValue;
+}
+
+void matchDereferenceExpr(const Stmt *stmt,
+                          const MatchFinder::MatchResult &Result,
+                          Environment &Env) {
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    // FIXME: Optimize
+    auto *RootValue = getValue(*Var, Env);
+    if (!RootValue) {
+       return;
+    }
+
+    Env.addToFlowCondition(Env.arena().makeNot(getVal(kIsNull, *RootValue).formula()));
+}
+
+void matchCastExpr(const CastExpr *cond, const MatchFinder::MatchResult &Result,
+                   ReverseNullModel::TransferArgs &Data) {
```

Comment:
🏗️ Here is something that feels a bit odd. So we have a Tidy checker in the works (same patch), which would result in matches to specific AST constructs, and we have this library implementation for the abstract data-flow model. However, now the model seems to want to use not only the AST elements (which would reasonably be well and alive in any sort of Clang-y context) but the specifics of the fact that you're driving the model **from** the checker.

So why is this model here and not in the anonymous namespace of the check's TU?

Conversely, if we want the model to be decoupled from the specifics of the check such that it is reusable from other libraries, then the actual relevant nodes should be given to this function like any other AST-based function, ignoring the match-result metastructure.

Reply:
Also, technically the model itself is decoupled from the checker - the only thing specific to the checker is the data layout within each boolean variable, and the `ReverseNullDiagnoser` class - the rest is just a generic reusable null-pointer analysis.